### PR TITLE
feat: add weeds and compost

### PR DESCRIPTION
**Pull request by lstebner**
note: this is a work in progress. the art is especially mediocre at this point but it's mostly working haha

### What this PR does

add "weeds" and "compost"

- weeds grow randomly overnight in unoccupied plots. they're worth very little but can be sold
- weeds can be harvested and turned into compost through crafting
- compost acts like fertilizer
    
### Additional information

Screen Shot 2022-07-17 at 9 18 21 PM
Screen Shot 2022-07-17 at 9 20 02 PM
Screen Shot 2022-07-17 at 9 20 16 PM
Screen Shot 2022-07-17 at 9 20 55 PM
Screen Shot 2022-07-17 at 9 23 59 PM
Screen Shot 2022-07-17 at 9 30 25 PM

### CODE REVIEW

**Comment by jeremyckahn** [comment]
This is very cool! I love the idea. I know you're still working on this so I won't dig into the code yet. I've got some initial play testing feedback (some of which you probably have in mind already):

- Weeds shouldn't be considered crops. I think it makes sense for weeds to just be a `farmhand.plotContent`, and not a `farmhand.crop`: https://github.com/jeremyckahn/farmhand/blob/a0384fdb029a35f0490f7f351e8bcac3cad6f173/src/index.js#L37-L44
  - As part of this, pulling weeds shouldn't trigger the `harvest-crop` achievement: https://github.com/jeremyckahn/farmhand/blob/a0384fdb029a35f0490f7f351e8bcac3cad6f173/src/data/achievements.js#L62-L69
- It might make sense to start new games with some weeds in the field.
- Crows shouldn't eat weeds.
- Weeds shouldn't have fluctuating prices. I think they should be sell-able for something like a stable $0.10.
- Perhaps instead of being directly plantable in the field, perhaps Compost should be an ingredient for a Fertilizer recipe?

File: src/game-logic/reducers/harvestPlot.js [hunk 1]
```diff
@@ -96,3 +85,40 @@ export const harvestPlot = (state, x, y) => {
...
+
+  if (!crop) return state
+
+  if (
+    getPlotContentType(crop) === itemType.CROP &&
+    getCropLifeStage(crop) === GROWN &&
+    doesInventorySpaceRemain(state)
+  ) {
+    return harvestCrops(state, x, y)
+  } else if (getPlotContentType(crop) === itemType.WEEDS) {
```
Comment (jeremyckahn) [inline src/game-logic/reducers/harvestPlot.js:119]: We'll want to check that `doesInventorySpaceRemain(state)` is `true` here as well so that weeds aren't pulled when there is no inventory space available to store them in.

File: src/game-logic/reducers/purchaseComposter.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+  if (purchasedComposter >= composterId) return state
+
+  state = {
+    ...state,
+    purchasedComposter: composterId,
+    money: moneyTotal(money, -PURCHASEABLE_COMPOSTERS.get(composterId).price),
+  }
+
+  //state = showNotification(state, COMPOSTER_AVAILABLE_NOTIFICATION)
```
Comment (jeremyckahn) [inline src/game-logic/reducers/purchaseComposter.js:25]: +1 to having a notification for this!

File: src/game-logic/reducers/spawnWeeds.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { WEEDS_SPAWN_CHANCE } from '../../constants'
+
+import { weeds } from '../../data/items'
+
+import isRandomNumberLessThan from '../../utils/isRandomNumberLessThan'
+import { getPlotContentFromItemId } from '../../utils'
+
+export function spawnWeeds(plotContents) {
```
Comment (jeremyckahn) [inline src/game-logic/reducers/spawnWeeds.js:8]: It would be good to have JSDoc annotations for this function.

File: src/enums.js [hunk 4]
```diff
@@ -29,6 +29,7 @@ export const cropType = enumify([
   'TOMATO',
   'WATERMELON',
   'WHEAT',
+  'WEEDS',
```
Comment (jeremyckahn) [inline src/enums.js:32]: What do you think about switching from plural "weeds" semantics to singular ("weed") for both the `cropType` here and the `itemType` below in order to remain consistent with the other `cropType` and `itemType` semantics?

File: src/enums.js [hunk 5]
```diff
@@ -90,13 +91,19 @@ export const itemType = enumify([
...
 /**
  * @property farmhand.module:enums.fertilizerType
  * @enum {string}
  */
-export const fertilizerType = enumify(['NONE', 'STANDARD', 'RAINBOW'])
+export const fertilizerType = enumify([
+  'NONE',
+  'STANDARD',
+  'RAINBOW',
+  'COMPOST',
```
Comment (jeremyckahn) [inline src/enums.js:105]: If we drop the mechanic of using compost as fertilizer directly, I think we can drop this addition.

File: src/components/Plot/Plot.js [hunk 6]
```diff
@@ -27,7 +27,11 @@ export const getBackgroundStyles = plotContent => {
 
   const backgroundImages = []
 
-  if (plotContent.fertilizerType === fertilizerType.STANDARD) {
+  if (
+    [fertilizerType.STANDARD, fertilizerType.COMPOST].includes(
```
Comment (jeremyckahn) [inline src/components/Plot/Plot.js:31]: If we drop `fertilizerType.COMPOST`, we can revert this change.

File: src/components/Inventory/Inventory.js [hunk 7]
```diff
@@ -61,6 +62,7 @@ const itemTypeCategoryMap = Object.freeze({
   [SPRINKLER]: FIELD_TOOLS,
   [STONE]: MINED_RESOURCES,
   [TOOL_UPGRADE]: UPGRADES,
+  [WEEDS]: CROPS,
```
Comment (jeremyckahn) [inline src/components/Inventory/Inventory.js:65]: Rather than categorizing Weeds under Crops, what do you think about introducing a new inventory category to display them under? Perhaps "Foraged Items?"

**Review by jeremyckahn — CHANGES_REQUESTED** [review]
Awesome work so far! We've already talked about a few things here on Discord, but I wanted to get it all in once place here on the PR. Hopefully none of my suggestions are too onerous to deal with.

The art looks great to me!

File: src/constants.js [hunk 8]
```diff
@@ -46,6 +46,10 @@ export const PURCHASEABLE_COMBINES = freeze(
   new Map([[1, { type: 'Basic', price: 500_000 }]])
 )
 
+export const PURCHASEABLE_COMPOSTERS = freeze(
+  new Map([[1, { type: 'Bin', price: 1_000 }]])
```
[the remark this answers is not in the document]
Comment (lstebner) [reply src/constants.js:50]: i thought it was a little bit more descriptive and it's what I own in real life, hah. i'm not married to it though and i'm not exactly sure what the upgrades would be called. "basic" is probably easier to work with as we introduce upgrades actually... i'll change it

File: src/game-logic/reducers/harvestPlot.js [hunk 9]
```diff
@@ -96,3 +85,40 @@ export const harvestPlot = (state, x, y) => {
     },
   }
 }
+
+function harvestWeeds(state, x, y) {
```
[the remark this answers is not in the document]
Comment (lstebner) [reply src/game-logic/reducers/harvestPlot.js:89]: if you feel strongly i can change it, but I like the "harvest" verbage here personally because it's part of the "harvestPlot" logic. and i kinda mentioned it in discord, but in my mind a more ultimate version is something that can actually call the same method on anything that is occupying the plot i.e. `plotItem.harvest()` rather than this reducer having to make a decision about how to handle what's in there. 

i am definitely going to singularize it though based on making the other stuff singular so it's "harvestWeed". like i said though, i can change it if you have a preference for "pull".

File: src/game-logic/reducers/harvestPlot.js (hunk 1 above)
Comment (lstebner) [reply src/game-logic/reducers/harvestPlot.js:119]: i pulled it out to be a guard clause since it applies to both scenarios

File: src/components/Inventory/Inventory.js (hunk 7 above)
Comment (lstebner) [reply src/components/Inventory/Inventory.js]: i really like this because it's a perfect spot for mushrooms to go too!

File: src/data/recipes.js [hunk 10]
```diff
@@ -392,3 +392,34 @@ export const goldIngot = itemify({
...
+
+export const compost = itemify({
+  id: 'compost',
+  name: 'Compost',
+  ingredients: {
+    [items.weeds.id]: 100,
+  },
+  condition: state =>
+    state.purchasedComposter && state.itemsSold[items.weeds.id] >= 100,
+  recipeType: recipeType.KITCHEN,
```
[the remark this answers is not in the document]
Comment (lstebner) [reply src/data/recipes.js]: done!

**Comment by lstebner** [comment]
thank you for the excellent and thorough review @jeremyckahn ! you had a ton of great suggestions and I believe I made them all (except one that I responded to). i also made an additional change to support multiple images for weeds and changed the graphic in general so make sure you give that code a look and let me know if it's ok or there's a better way to do it. after playing a bit i actually kinda like the compost image, but am open to modifying it if you have ideas.

File: src/components/Workshop/RecyclingTabPanel.js [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+import ReactMarkdown from 'react-markdown'
+
+import { recipeType } from '../../enums'
+
+import { recipeCategories } from '../../data/maps'
+
+import { TabPanel } from './TabPanel'
+import { RecipeList } from './RecipeList'
+
+export function RecyclingTabPanel({
```
Comment (jeremyckahn) [inline src/components/Workshop/RecyclingTabPanel.js:16]: Nice! 😁

File: src/data/recipes.js [hunk 12]
```diff
@@ -392,3 +392,34 @@ export const goldIngot = itemify({
     state.purchasedSmelter && state.itemsSold[items.goldOre.id] >= 50,
   recipeType: recipeType.FORGE,
 })
+
+export const compost = itemify({
```
Comment (jeremyckahn) [inline src/data/recipes.js:396]: The purpose of this item may not be obvious to players, so we might want to add a `description` explaining what it's for.

File: src/game-logic/reducers/harvestPlot.js [hunk 13]
```diff
@@ -96,3 +96,46 @@ export const harvestPlot = (state, x, y) => {
...
+  return state
+}
+
+/**
+ * @param {farmhand.state} state
+ * @param {number} x
+ * @param {number} y
+ * @returns {farmhand.state}
+ */
+export const harvestPlot = (state, x, y) => {
```
Comment (jeremyckahn) [inline src/game-logic/reducers/harvestPlot.js:124]: The new changes in this file make me really glad we moved each reducer to its own module!

File: src/game-logic/reducers/harvestPlot.js [hunk 14]
```diff
@@ -96,3 +96,46 @@ export const harvestPlot = (state, x, y) => {
...
+  const row = state.field[y]
+  const crop = row[x]
+
+  if (!crop) return state
+  if (!doesInventorySpaceRemain(state)) return state
```
Comment (jeremyckahn) [inline src/game-logic/reducers/harvestPlot.js:129]: :100:

File: src/utils.js [hunk 15]
```diff
@@ -376,7 +376,12 @@ export const getPlotImage = plotContent => {
       }
     }
 
-    if (plotContent?.oreId) {
+    if (getPlotContentType(plotContent) === itemType.WEED) {
+      const weedColors = ['yellow', 'orange', 'pink']
+      const color = weedColors[(x * y) % weedColors.length]
```
Comment (jeremyckahn) [inline src/utils.js:381]: This is a very smart way to determine the weed color! 👏

File: src/components/Inventory/Inventory.js (hunk 7 above)
Comment (jeremyckahn) [reply src/components/Inventory/Inventory.js]: Ooooh that would be an amazing addition. I love the idea! 🍄

File: src/game-logic/reducers/harvestPlot.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (jeremyckahn) [reply src/game-logic/reducers/harvestPlot.js]: I don't feel strongly about "pull" over "harvest," and I agree that it sets the stage for more scalable and generic semantics. Good thinking!

**Review by jeremyckahn — CHANGES_REQUESTED** [review]
This is looking great! I just have a few more (hopefully minor) comments, but we're almost there. I haven't had a chance to give this a thorough play-through since my last review, but I will do so just as soon as I have time.

Thanks for all the work and changes so far!

File: src/game-logic/reducers/purchaseComposter.js [hunk 16]
```diff
@@ -0,0 +1,28 @@
+import { moneyTotal } from '../../utils'
+import { PURCHASEABLE_COMPOSTERS } from '../../constants'
+import { RECYCLING_AVAILABLE_NOTIFICATION } from '../../strings'
+
+import { showNotification } from './showNotification'
+import { updateLearnedRecipes } from './updateLearnedRecipes'
+
+// TODO: Add tests for this reducer
```
[the remark this answers is not in the document]
Comment (lstebner) [reply src/game-logic/reducers/purchaseComposter.js:8]: i copied this from purchaseSmelter 😅 but it's a good callout. i'm also adding those tests to this PR

File: src/game-logic/reducers/harvestPlot.test.js [hunk 17]
```diff
@@ -34,6 +35,7 @@ describe('harvestPlot', () => {
       const inputState = {
         cropsHarvested: {},
         field: [[testCrop({ itemId: 'sample-crop-1' })]],
+        inventory: [],
```
[the remark this answers is not in the document]
Comment (lstebner) [reply src/game-logic/reducers/harvestPlot.test.js:39]: updated!

File: src/components/Workshop/RecyclingTabPanel.js (hunk 11 above)
Comment (lstebner) [reply src/components/Workshop/RecyclingTabPanel.js:16]: this was so easy thanks to the panel refactors done recently. very glad i did that 😄

File: src/img/index.js [hunk 18]
```diff
@@ -215,6 +222,10 @@ export const items = {
   watermelon,
   'watermelon-growing': watermelonGrowing,
   'watermelon-seed': watermelonSeed,
+  weed: weedYellow, // this one is used for the inventory
+  'weed-yellow': weedYellow,
+  'weed-orange': weedOrange,
+  'weed-pink': weedPink,
```
[the remark this answers is not in the document]
Comment (lstebner) [reply src/img/index.js:228]: changed to use the pink one

File: src/game-logic/reducers/purchaseSmelter.js [hunk 19]
```diff
@@ -5,7 +5,6 @@ import { FORGE_AVAILABLE_NOTIFICATION } from '../../strings'
 import { showNotification } from './showNotification'
 import { updateLearnedRecipes } from './updateLearnedRecipes'
 
-// TODO: Add tests for this reducer
```
Comment (lstebner) [inline src/game-logic/reducers/purchaseSmelter.js:8]: i had copied this file as my starting point for purchaseComposter so i went back and added the specs for this as well

**Comment by lstebner** [comment]
test coverage should be in a good place now so kickin this back your way @jeremyckahn

**Review by jeremyckahn — COMMENTED** [review]
The code looks great! Thanks for all the work you've done here. I did some playtesting and noticed some odd behavior when ending days while focused on the Field: 

[Screencast from 07-18-2022 09:47:23 PM.webm](https://user-images.githubusercontent.com/366330/179653776-d3fe0bde-2e19-4352-a813-3978ff90a58f.webm)

If you look closely, some of the weeds look like they're animating away when the day increments. They don't seem to be going into the player inventory, so I'm not sure what's happening here. Are you able to see this on your end?

**Comment by jeremyckahn** [comment]
> If you look closely, some of the weeds look like they're animating away when the day increments. They don't seem to be going into the player inventory, so I'm not sure what's happening here. Are you able to see this on your end?

I looked at the Field data and it looks like some plots with weeds just stop showing their graphic after a while (and are even animating away): 
Screenshot from 2022-07-18 22-04-51

I need to log off tonight but I will try to help debug this tomorrow!

**Comment by jeremyckahn** [comment]
One last thing: I think these parts of the code are related to this weird behavior:

https://github.com/jeremyckahn/farmhand/blob/3235c17120fc8ca9cd62f65b15817974e648c6d9/src/components/Plot/Plot.js#L81-L82

https://github.com/jeremyckahn/farmhand/blob/3235c17120fc8ca9cd62f65b15817974e648c6d9/src/components/Plot/Plot.js#L94-L102

I'm not sure how this all fits together, but the CSS class names these bits of logic set (`.was-just-shoveled` and `.is-ripe`) seem to be getting rendered when they shouldn't be... :thinking:

image

Save data I'm experiencing this in: [farmhand-2022-07-19.zip](https://github.com/jeremyckahn/farmhand/files/9137138/farmhand-2022-07-19.zip)

File: src/game-logic/reducers/purchaseSmelter.test.js [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+
+    test('it sets the purchased smelter', () => {
+      expect(newState.purchasedSmelter).toEqual(1)
+    })
+
+    test('it deducts the smelter cost', () => {
+      expect(newState.money).toEqual(0)
+    })
+
+    test('it shows the recycling available notification', () => {
```
[the remark this answers is not in the document]
Comment (lstebner) [reply src/game-logic/reducers/purchaseSmelter.test.js:31]: oh crap, you caught me 😅

**Comment by jeremyckahn** [comment]
I found a separate issue that may also be related to shoveling:
Screenshot from 2022-07-19 09-38-20

To reproduce this issue, load [crashing-farmhand-data.zip](https://github.com/jeremyckahn/farmhand/files/9142083/crashing-farmhand-data.zip), go to the Field, select the Hoe, and click on the field. This may or may not be related to shoveling some days prior, but I haven't been able to debug much yet.

**Comment by jeremyckahn** [comment]
https://github.com/jeremyckahn/farmhand/commit/99fdcaaa1be0ae19e83a756718f4f35d0a4ed41c seems to be addressing the [weed animation issue](https://github.com/jeremyckahn/farmhand/pull/311#issuecomment-1188548772)! Nice work on that. I'm still able to reproduce [the crashing issue](https://github.com/jeremyckahn/farmhand/pull/311#issuecomment-1188548772), but that seems like the last remaining bug.

**Comment by jeremyckahn** [comment]
> I'm still able to reproduce [the crashing issue](https://github.com/jeremyckahn/farmhand/pull/311#issuecomment-1188548772), but that seems like the last remaining bug.

This should be fixed in https://github.com/jeremyckahn/farmhand/commit/760bcb4001f93abafba2f524e5f0877cfaf42148! Please review and make sure it's working correctly for you.

I noticed some tests are failing due to https://github.com/jeremyckahn/farmhand/commit/99fdcaaa1be0ae19e83a756718f4f35d0a4ed41c, so we'll just need to determine whether these are legitimate breakages or just outdated tests.

**Comment by lstebner** [comment]
> This should be fixed in https://github.com/jeremyckahn/farmhand/commit/760bcb4001f93abafba2f524e5f0877cfaf42148! Please review and make sure it's working correctly for you.

thank you! i tested with the save you provided and it doesn't crash anymore 🎉 

> I noticed some tests are failing due to https://github.com/jeremyckahn/farmhand/commit/99fdcaaa1be0ae19e83a756718f4f35d0a4ed41c, so we'll just need to determine whether these are legitimate breakages or just outdated tests.

these should be fixed now!

File: src/components/Plot/Plot.js [hunk 21]
```diff
@@ -74,35 +74,20 @@ export const Plot = ({
...
-
-  useEffect(() => {
-    if (
-      !initialIsShoveledState &&
-      plotContent?.isShoveled &&
-      plotContent?.oreId
-    ) {
-      setWasJustShoveled(true)
-    }
-  }, [initialIsShoveledState, plotContent])
```
Comment (jeremyckahn) [inline src/components/Plot/Plot.js:101]: I think I know why this was here now. I found a newly-introduced bug wherein the "mined" animation replays when you leave the Field and then come back:

[Screencast from 07-20-2022 09:19:21 PM.webm](https://user-images.githubusercontent.com/366330/180116267-e9e7d6dd-721f-4ce1-a2af-5bd0e025ec59.webm)

We'll have to figure out a solution for this before we can merge. :confused:

**Review by jeremyckahn — CHANGES_REQUESTED** [review]
I found another bug. 😩

Unfortunately the `Plot` code is a little fragile, as we're seeing now. We are getting close though!

File: src/utils.js [hunk 22]
```diff
@@ -362,7 +362,7 @@ const cropLifeStageToImageSuffixMap = {
  * @param {farmhand.plotContent} plotContent
  * @returns {?string}
  */
-export const getPlotImage = plotContent => {
+export const getPlotImage = (plotContent, x, y) => {
```
[the remark this answers is not in the document]
Comment (jeremyckahn) [reply src/utils.js:367]: This is fixed in https://github.com/jeremyckahn/farmhand/commit/e7908c42451a6e3e73da97b112056e863a749368.

File: src/game-logic/reducers/resetWasShoveled.js [hunk 23]
```diff
@@ -7,6 +7,7 @@ export const resetWasShoveled = plotContent => {
     return {
       ...plotContent,
       daysUntilClear: plotContent.daysUntilClear - 1,
+      wasJustShoveled: false,
```
[the remark this answers is not in the document]
Comment (jeremyckahn) [reply src/game-logic/reducers/resetWasShoveled.js:10]: This is fixed in https://github.com/jeremyckahn/farmhand/commit/e7908c42451a6e3e73da97b112056e863a749368.

**Review by jeremyckahn — APPROVED** [review]
I think this is ready to go now! I just pushed some cleanup and refactoring work, so please review that and let me know if it looks good to you. If it does, feel free to merge and run a **minor** release since this change affects the persisted data structure.

Awesome work here! Thanks for sticking with this and knocking out all the bugs. And congrats on implementing a brand new feature! 🎉
